heareval/tasks/pipeline.py

```diff
@@ -0,0 +1,202 @@
+"""
+Generic pipelines for datasets
+"""
+
+from pathlib import Path
+import os
+from typing import List, Union
+from urllib.parse import urlparse
+
+import luigi
+
+from heareval.tasks.dataset_config import DatasetConfig
+import heareval.tasks.util.luigi as luigi_util
+
+
+def get_download_and_extract_tasks(config: DatasetConfig):
+
+    tasks = {}
+    for name, url in config["download_urls"].items():
+        filename = os.path.basename(urlparse(url).path)
+        task = luigi_util.ExtractArchive(
+            download=luigi_util.DownloadCorpus(
+                url=url, outfile=filename, data_config=config
+            ),
+            infile=filename,
+            outdir=name,
+            data_config=config,
+        )
+        tasks[name] = task
+
+    return tasks
+
+
+class SubsamplePartition(luigi_util.SubsamplePartition):
+    """
+    A subsampler that acts on a specific partition.
+    All instances of this will depend on the combined process metadata csv.
+    """
+
+    metadata = luigi.TaskParameter()
+
+    def requires(self):
+        # The meta files contain the path of the files in the data
+        # so we dont need to pass the extract as a dependency here.
+        return {
+            "meta": self.metadata,
+        }
+
+
+class SubsamplePartitions(luigi_util.WorkTask):
+    """
+    Aggregates subsampling of all the partitions into a single task as dependencies.
+    All the subsampled files are stored in the requires workdir, so we just link to
+    that since there aren't any real outputs associated with this task.
+    This is a bit of a hack -- but it allows us to avoid rewriting
+    the Subsample task as well as take advantage of Luigi concurrency.
+    """
+
+    metadata = luigi.TaskParameter()
+
+    def requires(self):
+        # Perform subsampling on each partition independently
+        subsample_partitions = {
+            partition["name"]: SubsamplePartition(
+                metadata=self.metadata,
+                partition=partition["name"],
+                max_files=partition["max_files"],
+                data_config=self.data_config,
+            )
+            for partition in self.data_config["partitions"]
+        }
+        return subsample_partitions
+
+    def run(self):
+        workdir = Path(self.workdir)
+        workdir.rmdir()
```

Comment:
Do we really need this rmdir? Can this cause bugs if some of the `SubsamplePartition` tasks complete successfully but then we run `SubsamplePartitions` again and it rmdirs?

Reply:
So the workdir in this task itself is a symlink to the previous requires workdir. The only time this is run is once all the  SubsamplePartition dependencies have completed - and it just creates a single symlink to the requires workdir (this basically just groups the multiple subsample dependencies into one). So the rmdir just removes a symlink if it exists, I think it is safe.